### ISSUE TITLE
Where is the adaptative average pooling?

### ISSUE THREAD

**Issue by mjack3** [amazon-research/patchcore-inspection#7]
Hello  snd thank you for this work.

Reading the paper, you say

"For PatchCore, we use adaptative average pooling..."
"with the optional use of a striding parameter s, which we set to 1 except for..."

I had looked for something like torch.nn.adatativePooling2D but I haven't found anything. Could you point the code where it happens or to explain what you mean with this average pooling? 

Thanks

**Comment by AlessandroPolidori** [comment]
@mjack3 i think it's in the MeanMapper class inside patchchore/common.py

**Comment by Confusezius** [comment]
Indeed - the pooling operation aims to aggregate the local neighbourhood surrounding each featuremap location. Conceptually, this is done using the `PatchMaker`-class 
https://github.com/amazon-research/patchcore-inspection/blob/b64be4734cb8295bfbadccf4f6a036b266181e57/src/patchcore/patchcore.py#L277-L308
and the `MeanMapper`
https://github.com/amazon-research/patchcore-inspection/blob/b64be4734cb8295bfbadccf4f6a036b266181e57/src/patchcore/common.py#L163-L170

The `PatchMaker` unfolds an input image batch `[bs x c x h x w]` into a patchified version `[bs x h//stride * w//stride x c x patchsize x patchsize]`. This is done for each featuremap that you extract from a given backbone network. 
The patchified featuremaps are then rescaled to match the highest available featuremap resolution, and then each patch is averaged using the `MeanMapper` class, which is our local neighbourhood aggregation.
Note that in this particular case, an adaptive average pooling could have also done the trick, but the patchified approach can allow for other methods of aggregation (although we did end up just opting for the simple averaging).

Everything I have mentioned happens here:
https://github.com/amazon-research/patchcore-inspection/blob/b64be4734cb8295bfbadccf4f6a036b266181e57/src/patchcore/patchcore.py#L105-L141


-------
Finally, just to highlight: the `Aggregator`-class
https://github.com/amazon-research/patchcore-inspection/blob/b64be4734cb8295bfbadccf4f6a036b266181e57/src/patchcore/common.py#L173-L183
also has an `adaptive_pooling` operation, but in this case only to bring the dimensionality of each utilized featuremap to a fixed `target_embed_dimension` - see also https://github.com/amazon-research/patchcore-inspection/issues/4#issuecomment-1123763551 for a more detailed explanation :).

**Comment by AlessandroPolidori** [comment]
@Confusezius So, correct me if i'm wrong, the paper's "aggregation function" part is done by the MeanMapper inside the "Preprocessing" step, right?

**Comment by Confusezius** [comment]
Indeed! And if you want to try different forms of aggregation, you can simply exchange the MeanMapper for another aggregation function on patches of size `patchsize x patchsize` :).

**Comment by mjack3** [comment]
Thanks for your answer. I'll check the comments as fast as I can. This is very interesting for me because I had worked in a PatchCore version where my adaptative pooling was something like 

```
self.local_smoothing = torch.nn.AdaptiveAvgPool2d((28, 28)) # 224 (input) // 8 (stride of layer 2)
self.locally_aware_patch = torch.nn.AvgPool2d(3, 1)
...

features = self.feature_extractor(x)
features[0] = self.local_smoothing(self.locally_aware_patch(features[0]))
features[1] = self.local_smoothing(self.locally_aware_patch(features[1]))

z = torch.cat(features, 1)
```

**Comment by Confusezius** [comment]
That seems like a sensible alternative! Feel free to update this issue with results/differences, I'm quite curious :).

**Comment by mjack3** [comment]
Sure, I 'll soon

**Comment by mjack3** [comment]
Well, here are the results using layer [2, 3], 1% of memory bank and 5 neighbour. I didn't read the latest paper version but my implementation seems just be worst, though, results changes a bit using several trials.

![Screenshot from 2022-05-24 14-34-58](https://user-images.githubusercontent.com/37028633/170035917-f8e2a0b3-6968-4cbe-b337-8f3f70df21bc.png)

By the way, I wonder why using a resolution biggest the performance can decrease. For example, using 224 seems to be better than using 448. Thinking as a human (hehe) you would see clearest the differences between bad and good. Moreover, this should be independent of the input used for training the backbone.

Just curious

**Comment by mjack3** [comment]
I am wondering what did lead you to decide to convert a tensor of dimension [bs x c x w x h] into a [bs * w//stride * h//stride, c, patchsize, patchsize] @Confusezius in the PatchMaker.

Kind regards

**Comment by wu33learn** [comment]
> Indeed! And if you want to try different forms of aggregation, you can simply exchange the MeanMapper for another aggregation function on patches of size `patchsize x patchsize` :).

hello,  i  am confused anbout the meanmaper, the input  : layer 2: (B*784, 1, 512*3*3)   layer3:  (B*784, 1, 1024*3*3),  meanmaper: F.adaptive_average_pooling(input,1024), it is right for the layer 3 form 1024*3*3 to 1024 that is local aggregation of neighboord with patchsize 3, while the layer2 , i am confused, 512*3*3  average pool to 1024, that seems unequal to the former, only about 5 in a patch 9 be aggregrated to average, and taht is computed twice to get two output in a neighborhood.  i am wondering if we can make 512*3*3 to 512 and then  interpolate to 1024, that makes use of the whole neighboreed information. looking forward to your reply！
